### PR TITLE
Replace withSelectedSite with useSelectedSiteSelector

### PR DESCRIPTION
**Pull request by lsl**
After some continued discussion in #57177 lets rename / replace the implementation with something else.

This is the change I'm proposing but I'm open to other ideas / naming.

#### Changes proposed in this Pull Request

* Replace `withSelectedSite` (#57177) with `useSelectedSiteSelector`.

#### Testing instructions

```
yarn run test-client client/state/sites/test/selectors.js
yarn run test-client client/state/sites/hooks/test/use-selected-site-selector.js 
```

Related to #57177

### CODE REVIEW

File: client/state/sites/selectors/index.js [hunk 1]
```diff
@@ -66,4 +66,4 @@ export { default as getSiteWooCommerceWizardUrl } from './get-site-woocommerce-w
 export { default as getSiteWordPressSeoWizardUrl } from './get-site-wordpress-seo-wizard-url';
 export { default as getJetpackSearchCustomizeUrl } from './get-jetpack-search-customize-url';
 export { default as getJetpackVersion } from './get-jetpack-version';
-export { default as withSelectedSite } from './with-selected-site';
+export { default as useSelectedSiteSelector } from './use-selected-site-selector';
```
Comment (jsnajdr) [inline client/state/sites/selectors/index.js]: Let's move it to `state/sites/hooks` and import from there. Nothing in `selectors/` should depend on the `react-redux` bindings.

**Comment by lsl** [comment]
@jsnajdr / @p-jackson Finally got back to this, I came across a use case for handling rest args so added that in, I think the previous suggestion confused me as it handled rest args in a different way because it was creating a function instead of just handing off. I think this works well / doesn't get too magic but curious about what you think.

**Comment by lsl** [comment]
Also lead me into this as a potential more generic version, but I'm not really sure how useful it is.
```
export type Selector< Params extends any[], Result > = (
	state: AppState,
	...args: Params
) => Result;

export function useSelectorWith< Params extends any[], Result >(
	selector: Selector< Params, Result >,
	...args: Params
): Result {
	return useSelector( ( state ) => selector( state, ...args ) );
}

// usage
const siteId = useSelector( getSelectedSiteId );
const subdomainAvailable = useSelectorWith( hasAvailableSiteFeature, siteId, FEATURE_WP_SUBDOMAIN );

```

File: client/state/sites/hooks/test/use-selected-site-selector.js [hunk 2]
```diff
@@ -0,0 +1,49 @@
+import { FEATURE_WP_SUBDOMAIN } from '@automattic/calypso-products';
+import * as redux from 'react-redux';
+import hasAvailableSiteFeature from 'calypso/state/selectors/has-available-site-feature';
+import { useSelectedSiteSelector } from '../';
+import { isJetpackSite } from '../../selectors';
+
+const useSelector = jest.spyOn( redux, 'useSelector' );
+
+describe( 'useSelectedSiteSelector()', () => {
+	test( 'returns selector with selected site added on call', () => {
```
Comment (lsl) [inline client/state/sites/hooks/test/use-selected-site-selector.js]:
```suggestion
	test( 'executes selector with selected site id passed in', () => {
```

File: client/state/sites/hooks/test/use-selected-site-selector.js [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+						jetpack: true,
+					},
+				},
+			},
+			ui: {
+				selectedSiteId: 123,
+			},
+		};
+		useSelector.mockImplementation( ( selector ) => selector( state ) );
+		expect( useSelectedSiteSelector( isJetpackSite ) ).toEqual( true );
```
Comment (jsnajdr) [inline client/state/sites/hooks/test/use-selected-site-selector.js]: nit: `.toBe( true )` is a bit better. `.toEqual` does a deep comparison of objects, and we don't need this extra power here.

```suggestion
		expect( useSelectedSiteSelector( isJetpackSite ) ).toBe( true );
```

**Comment by jsnajdr** [comment]
> `useSelectorWith( hasAvailableSiteFeature, siteId, FEATURE_WP_SUBDOMAIN )`

It saves just a few keystrokes, replacing:
```js
useSelector( state => _( state, # ) )
```
with:
```js
useSelectorWith( _, # );
```
I think it's not worth having to learn a new hook.

**Review by jsnajdr — APPROVED** [review]
:+1:

File: client/state/sites/hooks/test/use-selected-site-selector.js (hunk 3 above)
Comment (lsl) [reply client/state/sites/hooks/test/use-selected-site-selector.js]: Hijacked your comment to add suggestion, thanks!

**Pull request merged**